### PR TITLE
SqlClient and EntityFrameworkCore instrumentation - support for db.statement for text queries

### LINKED ISSUE #2331: Instrumentation.EntityFrameworkCore - possibility to configure `SetDbStatementForText` and `SetDbStatementForStoredProcedure`
Hi team,
I have setup and configuration of auto instrumentation for dotnet apps in container and integrated with Jaeger deployed to EKS Cluster,
the traces been show in Jaeger query, but didnt see the db statement call query.
Our apps using DB MySQL 8 and using dotnet

### PR DESCRIPTION
**Pull request by Kielek**
## Why

Fixes #2331

## What

SqlClient and EntityFrameworkCore instrumentation - support for db.statement for text queries by `OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT` and `OTEL_DOTNET_AUTO_ENTITYFRAMEWORKCORE_SET_DBSTATEMENT_FOR_TEXT`.

It is high demand feature in Splunk distribution. It will be great to include it on upstream side. I suppose that it will be beneficial also for another users.

## Tests

CI + extended tests.

## Checklist

- [x] `CHANGELOG.md` is updated.
- [x] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,12 @@ This component adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.h
 
 ### Added
 
+- Support `OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT`
+  and `OTEL_DOTNET_AUTO_ENTITYFRAMEWORKCORE_SET_DBSTATEMENT_FOR_TEXT`
+  (default values: `false`) environment variables which controls whether
+ `db.statement` attribute is set for text queries for SQL Client
+  and Entity Framework Core instrumentations.
```
Comment (theletterf) [inline CHANGELOG.md:16]:
```suggestion
- Support for `OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT`
  and `OTEL_DOTNET_AUTO_ENTITYFRAMEWORKCORE_SET_DBSTATEMENT_FOR_TEXT`
  environment variables which controls whether the `db.statement` attribute is set for text
  queries for SQL Client and Entity Framework Core instrumentations. The default value for
  both environment variables is `false` due to the risk of leaking sensitive information in the
  collected database queries.
```

File: test/IntegrationTests/EntityFrameworkCoreTests.cs [hunk 2]
```diff
@@ -15,18 +16,43 @@ public EntityFrameworkCoreTests(ITestOutputHelper output)
...
-    [MemberData(nameof(LibraryVersion.EntityFrameworkCore), MemberType = typeof(LibraryVersion))]
-    public void SubmitsTraces(string packageVersion)
+    [MemberData(nameof(GetData))]
+    public void SubmitTraces(string packageVersion, bool dbStatementForText)
     {
+        SetEnvironmentVariable("OTEL_DOTNET_AUTO_ENTITYFRAMEWORKCORE_SET_DBSTATEMENT_FOR_TEXT", dbStatementForText.ToString());
         using var collector = new MockSpansCollector(Output);
         SetExporter(collector);
-        collector.Expect("OpenTelemetry.Instrumentation.EntityFrameworkCore");
+        collector.Expect("OpenTelemetry.Instrumentation.EntityFrameworkCore", span => Expect(span, dbStatementForText));
```
Comment (pellared) [inline test/IntegrationTests/EntityFrameworkCoreTests.cs:36]: I find `private static bool Expect(Span span, bool dbStatementForText)` too complex and I also do not like the name. I think having the logic inline is easier to follow.

```suggestion
        if (dbStatementForText)
        {
            collector.Expect("OpenTelemetry.Instrumentation.EntityFrameworkCore", span => span.Attributes.Any(attr => attr.Key == "db.statement" && !string.IsNullOrWhiteSpace(attr.Value?.StringValue)));
        }
        else
            collector.Expect("OpenTelemetry.Instrumentation.EntityFrameworkCore", span => !span.Attributes.Any(attr => attr.Key == "db.statement"));
        }
```

Comment (Kielek) [reply test/IntegrationTests/EntityFrameworkCoreTests.cs]: 6a0fea137be9334f77449cd3d7c0cf36d5ef2f79

**Review by RassK — APPROVED** [review]

**Review by pellared — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs [hunk 3]
```diff
@@ -122,6 +122,18 @@ public static class InstrumentationOptions
             /// Configuration key for GraphQL instrumentation to enable passing query as a document attribute.
             /// </summary>
             public const string GraphQLSetDocument = "OTEL_DOTNET_AUTO_GRAPHQL_SET_DOCUMENT";
+
+            /// <summary>
+            /// Configuration key for SQL Client instrumentation to enable passing text query as a db.statement attribute.
+            /// </summary>
+            public const string SqlClientSetDbStatementForText = "OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT";
```
Comment (lachmatt) [inline src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs:129]: For the sake of clarity, I'd change `OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT` to `OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT_QUERIES` (i.e add `_QUERIES` suffix)

Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs]: I like the change, but it is a bit different than in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2331#issuecomment-1926638418.

 Fixed in ed0fd0de552bed1026105107116efb0042635e44.

File: docs/config.md [hunk 4]
```diff
@@ -198,9 +198,11 @@ the `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` environment variable to
...
+| Environment variable                                            | Description                                                                                                                                                | Default value | Status                                                                                                                            |
+|-----------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------|-----------------------------------------------------------------------------------------------------------------------------------|
+| `OTEL_DOTNET_AUTO_ENTITYFRAMEWORKCORE_SET_DBSTATEMENT_FOR_TEXT` | Whether the Entity Framework Core instrumentation can pass text queries through the `db.statement` attribute. Queries might contain sensitive information. | `false`       | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
```
Comment (pellared) [inline docs/config.md]: This is not only for "text queries" but for all "SQL statements".  We should tell that if it is `false` (default) then only executions of  stored procedures pass a value to `db.statement`.

**Review by pellared — CHANGES_REQUESTED** [review]
Second thought after a chat with @lachmatt . The description should be more clear and user-facing.

File: src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs (hunk 3 above)
Comment (lachmatt) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs]: I realized instrumentation packages use `FOR_TEXT` suffix in properties names, I'm fine with initial names.

Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs]: I disagree as this is not only for queries.

Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/ConfigurationKeys.cs:129]: @pellared , see https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs#L190-L215

There are 3 types:
 - stored procedures
 - text
 - table direct

We are touching here text.

Same for SQL Client.

**Review by pellared — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,13 @@ This component adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.h
 
 ### Added
 
+- Support for `OTEL_DOTNET_AUTO_SQLCLIENT_SET_DBSTATEMENT_FOR_TEXT`
+  and `OTEL_DOTNET_AUTO_ENTITYFRAMEWORKCORE_SET_DBSTATEMENT_FOR_TEXT`
+  environment variables which controls whether the `db.statement`
+  attribute is set for text  queries for SQL Client and Entity Framework Core
```
Comment (pellared) [inline CHANGELOG.md:15]:
```suggestion
  attribute is set for SQL statements for SQL Client and Entity Framework Core
```

**Review by rajkumar-rangaraj — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]

**Pull request merged**